src/catalog/database_catalog.cpp

```diff
@@ -3,35 +3,419 @@
+  // Step 1: Read Index
+  std::vector<col_oid_t> table_oids{ATTNUM_COL_OID, ATTNAME_COL_OID,    ATTTYPID_COL_OID,
+                                    ATTLEN_COL_OID, ATTNOTNULL_COL_OID, ADBIN_COL_OID};
+  // NOLINTNEXTLINE
+  auto [table_pri, table_pm] = columns_->InitializerForProjectedRow(table_oids);
+  auto oid_pri = columns_oid_index_->GetProjectedRowInitializer();
+  // Buffer is large enough to hold all prs
+  byte *buffer = common::AllocationUtil::AllocateAligned(table_pri.ProjectedRowSize());
+  // Scan the oid index
+  auto pr = oid_pri.InitializeRow(buffer);
+  auto *oid_entry = reinterpret_cast<uint32_t *>(pr->AccessForceNotNull(0));
+  *oid_entry = col_oid;
+  auto *relid_entry = reinterpret_cast<uint32_t *>(pr->AccessForceNotNull(1));
+  *relid_entry = class_oid;
+  std::vector<storage::TupleSlot> index_results;
+  columns_oid_index_->ScanKey(*txn, *pr, &index_results);
+  if (index_results.empty()) {
+    delete[] buffer;
+    return nullptr;
+  }
+  TERRIER_ASSERT(index_results.size() == 1, "Columns oid not unique in index");
+
+  // Step 2: Scan the table to get the column
+  pr = table_pri.InitializeRow(buffer);
+  if (!columns_->Select(txn, index_results[0], pr)) {
+    // Nothing visible
+    delete[] buffer;
+    return nullptr;
+  }
+  auto col = postgres::AttributeHelper::MakeColumn<Column>(pr, table_pm);
+  delete[] buffer;
+  return col;
+}
+
+template <typename Column>
+std::vector<std::unique_ptr<Column>> DatabaseCatalog::GetTableAttributes(transaction::TransactionContext *txn,
+                                                                         uint32_t class_oid) {
+  // Step 1: Read Index
+  std::vector<col_oid_t> table_oids{ATTNUM_COL_OID, ATTNAME_COL_OID,    ATTTYPID_COL_OID,
+                                    ATTLEN_COL_OID, ATTNOTNULL_COL_OID, ADBIN_COL_OID};
+  // NOLINTNEXTLINE
+  auto [table_pri, table_pm] = columns_->InitializerForProjectedRow(table_oids);
+  auto class_pri = columns_class_index_->GetProjectedRowInitializer();
+  // Buffer is large enough to hold all prs
+  byte *buffer = common::AllocationUtil::AllocateAligned(table_pri.ProjectedRowSize());
+  // Scan the class index
+  auto pr = class_pri.InitializeRow(buffer);
+  auto *relid_entry = reinterpret_cast<uint32_t *>(pr->AccessForceNotNull(1));
+  *relid_entry = class_oid;
+  std::vector<storage::TupleSlot> index_results;
+  columns_class_index_->ScanKey(*txn, *pr, &index_results);
+  if (index_results.empty()) {
+    delete[] buffer;
+    return nullptr;
+  }
+
+  // Step 2: Scan the table to get the columns
+  std::vector<std::unique_ptr<Column>> cols;
+  pr = table_pri.InitializeRow(buffer);
+  for (const auto &slot : index_results) {
+    if (!columns_->Select(txn, slot, pr)) {
+      // Nothing visible
+      delete[] buffer;
+      return nullptr;
+    }
+    cols.emplace_back(postgres::AttributeHelper::MakeColumn<Column>(pr, table_pm));
+  }
+  delete[] buffer;
+  return cols;
+}
+
+template <typename Column>
```

Comment:
Does this one need to be templated?

Reply:
It's used inside the function to read a column and delete it from the various indexes.